modules/fedimint-ln-client/src/lib.rs

```diff
@@ -1072,6 +1087,71 @@ impl LightningClientModule {
         ))
     }
 
+    async fn select_gateway_with_policy<T, E, F, Fut>(
+        &self,
+        maybe_gateway: Option<LightningGateway>,
+        operation: F,
+    ) -> anyhow::Result<T, E>
+    where
+        F: Fn(LightningGateway) -> Fut,
+        Fut: Future<Output = anyhow::Result<T, E>>,
+        E: std::fmt::Debug + From<anyhow::Error> + std::error::Error + Send + Sync + 'static,
+    {
+        if let Some(gw) = maybe_gateway {
+            let gw_id = gw.gateway_id;
+            return operation(gw.clone())
+                .await
+                .with_context(|| format!("Failed with specified gateway: {}", gw_id))
+                .map_err(E::from);
+        }
+
+        let gateways = self.list_gateways().await;
+        if gateways.is_empty() {
+            return Err(E::from(anyhow!("No gateways available")));
+        }
+
+        let gateways_with_status =
+            futures::future::join_all(gateways.into_iter().map(|gw| async {
+                let online = self
+                    .gateway_conn
+                    .verify_gateway_availability(&gw.info)
+                    .await
+                    .is_ok();
+                (gw, online)
+            }))
+            .await;
+
+        let sorted_gateways = gateways_with_status
+            .into_iter()
+            .sorted_by(|(a, a_online), (b, b_online)| {
+                let a_priority = match (a_online, a.vetted) {
+                    (true, true) => 0,   // Online+Vetted
+                    (true, false) => 1,  // Online
+                    (false, true) => 2,  // Offline+Vetted
+                    (false, false) => 3, // Offline
+                };
+                let b_priority = match (b_online, b.vetted) {
+                    (true, true) => 0,
+                    (true, false) => 1,
+                    (false, true) => 2,
+                    (false, false) => 3,
+                };
+                a_priority.cmp(&b_priority)
+            })
+            .collect::<Vec<_>>();
+
+        for (gw, _online) in sorted_gateways {
+            match operation(gw.info.clone()).await {
+                Ok(result) => return Ok(result),
+                Err(_err) => {
+                    continue;
+                }
+            }
+        }
```

Comment:
It's unclear to me if we should even retry an operation that failed with a gw that we consider online. I'll defer to @m1sterc001guy though